src/Functions/FunctionBinaryArithmetic.h

```diff
@@ -894,9 +894,8 @@ class FunctionBinaryArithmetic : public IFunction
             const NativeResultType const_b = helperGetOrConvert<T1, ResultDataType>(col_right_const, right);
 
             const ResultType res = check_decimal_overflow
-                // the arguments are already scaled after conversion
-                ? OpImplCheck::template process<left_is_decimal, right_is_decimal>(const_a, const_b, 1, 1)
-                : OpImpl::template process<left_is_decimal, right_is_decimal>(const_a, const_b, 1, 1);
+                ? OpImplCheck::template process<left_is_decimal, right_is_decimal>(const_a, const_b, scale_a, scale_b)
+                : OpImpl::template process<left_is_decimal, right_is_decimal>(const_a, const_b, scale_a, scale_b);
 
             if constexpr (result_is_decimal)
                 return ResultDataType(type.getPrecision(), type.getScale()).createColumnConst(
```

Comment:
Not sure whether it's the right fix or not, while developing the templated solution I've encountered various bugs due to the fact the `const_float * const_decimal` and vice verca __are__ scaled in `helperGetOrConvert`, so when the `scale_a` and `scale_b` weren't 1, the `process` function produced over- and under-scaling.

The general problem here is that for different operations (multiply, divide, etc...) the process function works differenly (and I couldn't change it while refactoring the initial code.)

Let's wait for the tests.

Reply:
As far as I tried, it is the right fix. Thanks